// arduino-ide-extension/lib/browser/contributions/contribution.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SketchContribution = exports.Contribution = exports.open = exports.Sketch = exports.URI = exports.TabBarToolbarRegistry = exports.KeybindingRegistry = exports.MenuModelRegistry = exports.CommandRegistry = exports.Command = void 0;
var inversify_1 = require("inversify");
var uri_1 = require("@theia/core/lib/common/uri");
exports.URI = uri_1.default;
var logger_1 = require("@theia/core/lib/common/logger");
var common_1 = require("@theia/filesystem/lib/common");
var label_provider_1 = require("@theia/core/lib/browser/label-provider");
var message_service_1 = require("@theia/core/lib/common/message-service");
var workspace_service_1 = require("@theia/workspace/lib/browser/workspace-service");
var opener_service_1 = require("@theia/core/lib/browser/opener-service");
Object.defineProperty(exports, "open", { enumerable: true, get: function () { return opener_service_1.open; } });
var menu_1 = require("@theia/core/lib/common/menu");
Object.defineProperty(exports, "MenuModelRegistry", { enumerable: true, get: function () { return menu_1.MenuModelRegistry; } });
var keybinding_1 = require("@theia/core/lib/browser/keybinding");
Object.defineProperty(exports, "KeybindingRegistry", { enumerable: true, get: function () { return keybinding_1.KeybindingRegistry; } });
var tab_bar_toolbar_1 = require("@theia/core/lib/browser/shell/tab-bar-toolbar");
Object.defineProperty(exports, "TabBarToolbarRegistry", { enumerable: true, get: function () { return tab_bar_toolbar_1.TabBarToolbarRegistry; } });
var command_1 = require("@theia/core/lib/common/command");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return command_1.Command; } });
Object.defineProperty(exports, "CommandRegistry", { enumerable: true, get: function () { return command_1.CommandRegistry; } });
var editor_mode_1 = require("../editor-mode");
var sketches_service_client_impl_1 = require("../../common/protocol/sketches-service-client-impl");
var protocol_1 = require("../../common/protocol");
Object.defineProperty(exports, "Sketch", { enumerable: true, get: function () { return protocol_1.Sketch; } });
var Contribution = /** @class */ (function () {
    function Contribution() {
    }
    Contribution.prototype.registerCommands = function (registry) {
    };
    Contribution.prototype.registerMenus = function (registry) {
    };
    Contribution.prototype.registerKeybindings = function (registry) {
    };
    Contribution.prototype.registerToolbarItems = function (registry) {
    };
    __decorate([
        inversify_1.inject(logger_1.ILogger),
        __metadata("design:type", Object)
    ], Contribution.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(message_service_1.MessageService),
        __metadata("design:type", message_service_1.MessageService)
    ], Contribution.prototype, "messageService", void 0);
    __decorate([
        inversify_1.inject(command_1.CommandService),
        __metadata("design:type", Object)
    ], Contribution.prototype, "commandService", void 0);
    __decorate([
        inversify_1.inject(workspace_service_1.WorkspaceService),
        __metadata("design:type", workspace_service_1.WorkspaceService)
    ], Contribution.prototype, "workspaceService", void 0);
    __decorate([
        inversify_1.inject(editor_mode_1.EditorMode),
        __metadata("design:type", editor_mode_1.EditorMode)
    ], Contribution.prototype, "editorMode", void 0);
    __decorate([
        inversify_1.inject(label_provider_1.LabelProvider),
        __metadata("design:type", label_provider_1.LabelProvider)
    ], Contribution.prototype, "labelProvider", void 0);
    Contribution = __decorate([
        inversify_1.injectable()
    ], Contribution);
    return Contribution;
}());
exports.Contribution = Contribution;
var SketchContribution = /** @class */ (function (_super) {
    __extends(SketchContribution, _super);
    function SketchContribution() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        inversify_1.inject(common_1.FileSystem),
        __metadata("design:type", Object)
    ], SketchContribution.prototype, "fileSystem", void 0);
    __decorate([
        inversify_1.inject(protocol_1.FileSystemExt),
        __metadata("design:type", Object)
    ], SketchContribution.prototype, "fileSystemExt", void 0);
    __decorate([
        inversify_1.inject(protocol_1.ConfigService),
        __metadata("design:type", Object)
    ], SketchContribution.prototype, "configService", void 0);
    __decorate([
        inversify_1.inject(protocol_1.SketchesService),
        __metadata("design:type", Object)
    ], SketchContribution.prototype, "sketchService", void 0);
    __decorate([
        inversify_1.inject(opener_service_1.OpenerService),
        __metadata("design:type", Object)
    ], SketchContribution.prototype, "openerService", void 0);
    __decorate([
        inversify_1.inject(sketches_service_client_impl_1.SketchesServiceClientImpl),
        __metadata("design:type", sketches_service_client_impl_1.SketchesServiceClientImpl)
    ], SketchContribution.prototype, "sketchServiceClient", void 0);
    SketchContribution = __decorate([
        inversify_1.injectable()
    ], SketchContribution);
    return SketchContribution;
}(Contribution));
exports.SketchContribution = SketchContribution;
(function (Contribution) {
    function configure(bind, serviceIdentifier) {
        bind(serviceIdentifier).toSelf().inSingletonScope();
        bind(command_1.CommandContribution).toService(serviceIdentifier);
        bind(menu_1.MenuContribution).toService(serviceIdentifier);
        bind(keybinding_1.KeybindingContribution).toService(serviceIdentifier);
        bind(tab_bar_toolbar_1.TabBarToolbarContribution).toService(serviceIdentifier);
    }
    Contribution.configure = configure;
})(Contribution = exports.Contribution || (exports.Contribution = {}));
exports.Contribution = Contribution;
