// arduino-ide-extension/lib/browser/contributions/verify-sketch.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifySketch = void 0;
var inversify_1 = require("inversify");
var output_channel_1 = require("@theia/output/lib/common/output-channel");
var protocol_1 = require("../../common/protocol");
var arduino_menus_1 = require("../menu/arduino-menus");
var arduino_toolbar_1 = require("../toolbar/arduino-toolbar");
var boards_data_store_1 = require("../boards/boards-data-store");
var boards_service_client_impl_1 = require("../boards/boards-service-client-impl");
var contribution_1 = require("./contribution");
var VerifySketch = /** @class */ (function (_super) {
    __extends(VerifySketch, _super);
    function VerifySketch() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    VerifySketch_1 = VerifySketch;
    VerifySketch.prototype.registerCommands = function (registry) {
        var _this = this;
        registry.registerCommand(VerifySketch_1.Commands.VERIFY_SKETCH, {
            execute: function () { return _this.verifySketch(); }
        });
        registry.registerCommand(VerifySketch_1.Commands.VERIFY_SKETCH_TOOLBAR, {
            isVisible: function (widget) { return arduino_toolbar_1.ArduinoToolbar.is(widget) && widget.side === 'left'; },
            execute: function () { return registry.executeCommand(VerifySketch_1.Commands.VERIFY_SKETCH.id); }
        });
    };
    VerifySketch.prototype.registerMenus = function (registry) {
        registry.registerMenuAction(arduino_menus_1.ArduinoMenus.SKETCH__MAIN_GROUP, {
            commandId: VerifySketch_1.Commands.VERIFY_SKETCH.id,
            label: 'Verify/Compile',
            order: '2'
        });
    };
    VerifySketch.prototype.registerKeybindings = function (registry) {
        registry.registerKeybinding({
            command: VerifySketch_1.Commands.VERIFY_SKETCH.id,
            keybinding: 'CtrlCmd+R'
        });
    };
    VerifySketch.prototype.registerToolbarItems = function (registry) {
        registry.registerItem({
            id: VerifySketch_1.Commands.VERIFY_SKETCH_TOOLBAR.id,
            command: VerifySketch_1.Commands.VERIFY_SKETCH_TOOLBAR.id,
            tooltip: 'Verify',
            priority: 0
        });
    };
    VerifySketch.prototype.verifySketch = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, boardsConfig, fqbn, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.sketchServiceClient.currentSketchFile()];
                    case 1:
                        uri = _a.sent();
                        if (!uri) {
                            return [2 /*return*/];
                        }
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 5, , 6]);
                        boardsConfig = this.boardsServiceClientImpl.boardsConfig;
                        if (!boardsConfig || !boardsConfig.selectedBoard) {
                            throw new Error('No boards selected. Please select a board.');
                        }
                        if (!boardsConfig.selectedBoard.fqbn) {
                            throw new Error("No core is installed for the '" + boardsConfig.selectedBoard.name + "' board. Please install the core.");
                        }
                        return [4 /*yield*/, this.boardsDataStore.appendConfigToFqbn(boardsConfig.selectedBoard.fqbn)];
                    case 3:
                        fqbn = _a.sent();
                        this.outputChannelManager.getChannel('Arduino: compile').clear();
                        return [4 /*yield*/, this.coreService.compile({
                                sketchUri: uri,
                                fqbn: fqbn,
                                optimizeForDebug: this.editorMode.compileForDebug
                            })];
                    case 4:
                        _a.sent();
                        this.messageService.info('Done compiling.', { timeout: 1000 });
                        return [3 /*break*/, 6];
                    case 5:
                        e_1 = _a.sent();
                        this.messageService.error(e_1.toString());
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    var VerifySketch_1;
    __decorate([
        inversify_1.inject(protocol_1.CoreService),
        __metadata("design:type", Object)
    ], VerifySketch.prototype, "coreService", void 0);
    __decorate([
        inversify_1.inject(boards_data_store_1.BoardsDataStore),
        __metadata("design:type", boards_data_store_1.BoardsDataStore)
    ], VerifySketch.prototype, "boardsDataStore", void 0);
    __decorate([
        inversify_1.inject(boards_service_client_impl_1.BoardsServiceClientImpl),
        __metadata("design:type", boards_service_client_impl_1.BoardsServiceClientImpl)
    ], VerifySketch.prototype, "boardsServiceClientImpl", void 0);
    __decorate([
        inversify_1.inject(output_channel_1.OutputChannelManager),
        __metadata("design:type", output_channel_1.OutputChannelManager)
    ], VerifySketch.prototype, "outputChannelManager", void 0);
    VerifySketch = VerifySketch_1 = __decorate([
        inversify_1.injectable()
    ], VerifySketch);
    return VerifySketch;
}(contribution_1.SketchContribution));
exports.VerifySketch = VerifySketch;
(function (VerifySketch) {
    var Commands;
    (function (Commands) {
        Commands.VERIFY_SKETCH = {
            id: 'arduino-verify-sketch'
        };
        Commands.VERIFY_SKETCH_TOOLBAR = {
            id: 'arduino-verify-sketch--toolbar'
        };
    })(Commands = VerifySketch.Commands || (VerifySketch.Commands = {}));
})(VerifySketch = exports.VerifySketch || (exports.VerifySketch = {}));
exports.VerifySketch = VerifySketch;
