// arduino-ide-extension/lib/electron-browser/theia/core/electron-menu-contribution.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElectronMenuContribution = void 0;
var inversify_1 = require("inversify");
var electron_menu_contribution_1 = require("@theia/core/lib/electron-browser/menu/electron-menu-contribution");
var ElectronMenuContribution = /** @class */ (function (_super) {
    __extends(ElectronMenuContribution, _super);
    function ElectronMenuContribution() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ElectronMenuContribution.prototype.hideTopPanel = function () {
        // NOOP
        // We reuse the `div` for the Arduino toolbar.
    };
    ElectronMenuContribution.prototype.update = function () {
        this.setMenu();
    };
    ElectronMenuContribution.prototype.registerCommands = function (registry) {
        _super.prototype.registerCommands.call(this, registry);
        registry.unregisterCommand(electron_menu_contribution_1.ElectronCommands.CLOSE_WINDOW);
    };
    ElectronMenuContribution.prototype.registerMenus = function (registry) {
        _super.prototype.registerMenus.call(this, registry);
        registry.unregisterMenuAction(electron_menu_contribution_1.ElectronCommands.CLOSE_WINDOW);
    };
    ElectronMenuContribution.prototype.registerKeybindings = function (registry) {
        _super.prototype.registerKeybindings.call(this, registry);
        registry.unregisterKeybinding(electron_menu_contribution_1.ElectronCommands.CLOSE_WINDOW.id);
        registry.unregisterKeybinding(electron_menu_contribution_1.ElectronCommands.ZOOM_IN.id);
        registry.unregisterKeybinding(electron_menu_contribution_1.ElectronCommands.ZOOM_OUT.id);
    };
    ElectronMenuContribution = __decorate([
        inversify_1.injectable()
    ], ElectronMenuContribution);
    return ElectronMenuContribution;
}(electron_menu_contribution_1.ElectronMenuContribution));
exports.ElectronMenuContribution = ElectronMenuContribution;
