// arduino-ide-extension/lib/browser/theia/editor/editor-contribution.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorContribution = void 0;
var inversify_1 = require("inversify");
var editor_contribution_1 = require("@theia/editor/lib/browser/editor-contribution");
var browser_1 = require("@theia/core/lib/browser");
var EditorContribution = /** @class */ (function (_super) {
    __extends(EditorContribution, _super);
    function EditorContribution() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EditorContribution.prototype.updateLanguageStatus = function (editor) {
    };
    EditorContribution.prototype.setCursorPositionStatus = function (editor) {
        if (!editor) {
            this.statusBar.removeElement('editor-status-cursor-position');
            return;
        }
        var cursor = editor.cursor;
        this.statusBar.setElement('editor-status-cursor-position', {
            text: "" + (cursor.line + 1),
            alignment: browser_1.StatusBarAlignment.LEFT,
            priority: 100
        });
    };
    EditorContribution = __decorate([
        inversify_1.injectable()
    ], EditorContribution);
    return EditorContribution;
}(editor_contribution_1.EditorContribution));
exports.EditorContribution = EditorContribution;
