// arduino-ide-extension/lib/node/fs-extra.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mkdirpSync = exports.mkdirp = exports.unwatchFile = exports.watchFile = exports.rename = exports.access = exports.readFile = exports.writeFile = exports.stat = exports.readdir = exports.lstat = exports.exists = exports.renameSync = exports.accessSync = exports.readFileSync = exports.writeFileSync = exports.statSync = exports.readdirSync = exports.lstatSync = exports.existsSync = exports.constants = void 0;
var fs = require("fs");
var util_1 = require("util");
exports.constants = fs.constants;
exports.existsSync = fs.existsSync;
exports.lstatSync = fs.lstatSync;
exports.readdirSync = fs.readdirSync;
exports.statSync = fs.statSync;
exports.writeFileSync = fs.writeFileSync;
exports.readFileSync = fs.readFileSync;
exports.accessSync = fs.accessSync;
exports.renameSync = fs.renameSync;
exports.exists = util_1.promisify(fs.exists);
exports.lstat = util_1.promisify(fs.lstat);
exports.readdir = util_1.promisify(fs.readdir);
exports.stat = util_1.promisify(fs.stat);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.readFile = util_1.promisify(fs.readFile);
exports.access = util_1.promisify(fs.access);
exports.rename = util_1.promisify(fs.rename);
exports.watchFile = fs.watchFile;
exports.unwatchFile = fs.unwatchFile;
function mkdirp(path, timeout) {
    if (timeout === void 0) { timeout = 3000; }
    return new Promise(function (resolve, reject) {
        var timeoutHandle;
        if (timeout > 0) {
            timeoutHandle = setTimeout(function () {
                reject(new Error("Timeout of " + timeout + " ms exceeded while trying to create the directory \"" + path + "\""));
            }, timeout);
        }
        fs.mkdir(path, { recursive: true }, function (err) {
            clearTimeout(timeoutHandle);
            if (err)
                reject(err);
            else
                resolve();
        });
    });
}
exports.mkdirp = mkdirp;
function mkdirpSync(path) {
    fs.mkdirSync(path, { recursive: true });
}
exports.mkdirpSync = mkdirpSync;
