// arduino-ide-extension/lib/browser/arduino-daemon-client-impl.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var logger_1 = require("@theia/core/lib/common/logger");
var event_1 = require("@theia/core/lib/common/event");
var message_service_1 = require("@theia/core/lib/common/message-service");
var ArduinoDaemonClientImpl = /** @class */ (function () {
    function ArduinoDaemonClientImpl() {
        this.onStartedEmitter = new event_1.Emitter();
        this.onStoppedEmitter = new event_1.Emitter();
        this._isRunning = false;
    }
    ArduinoDaemonClientImpl.prototype.notifyStopped = function () {
        if (this._isRunning) {
            this._isRunning = false;
            this.onStoppedEmitter.fire();
            this.info('The CLI daemon process has stopped.');
        }
    };
    ArduinoDaemonClientImpl.prototype.notifyStarted = function () {
        if (!this._isRunning) {
            this._isRunning = true;
            this.onStartedEmitter.fire();
            this.info('The CLI daemon process has started.');
        }
    };
    Object.defineProperty(ArduinoDaemonClientImpl.prototype, "onDaemonStarted", {
        get: function () {
            return this.onStartedEmitter.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ArduinoDaemonClientImpl.prototype, "onDaemonStopped", {
        get: function () {
            return this.onStoppedEmitter.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ArduinoDaemonClientImpl.prototype, "isRunning", {
        get: function () {
            return this._isRunning;
        },
        enumerable: true,
        configurable: true
    });
    ArduinoDaemonClientImpl.prototype.info = function (message) {
        this.messageService.info(message, { timeout: 3000 });
        this.logger.info(message);
    };
    __decorate([
        inversify_1.inject(logger_1.ILogger),
        __metadata("design:type", Object)
    ], ArduinoDaemonClientImpl.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(message_service_1.MessageService),
        __metadata("design:type", message_service_1.MessageService)
    ], ArduinoDaemonClientImpl.prototype, "messageService", void 0);
    ArduinoDaemonClientImpl = __decorate([
        inversify_1.injectable()
    ], ArduinoDaemonClientImpl);
    return ArduinoDaemonClientImpl;
}());
exports.ArduinoDaemonClientImpl = ArduinoDaemonClientImpl;
